modules/core/src/main/scala/sangria/execution/ValueCoercionHelper.scala

```diff
@@ -580,107 +580,126 @@ class ValueCoercionHelper[Ctx](
+      Vector(violation)
+    }
 
-      val fieldViolations =
-        objTpe.fields.toVector.flatMap(f =>
-          isValidValue(f.fieldType, um.getMapValue(valueMap, f.name))
-            .map(MapValueViolation(f.name, _, sourceMapper, Nil)))
+    def isValidValueRec(tpe: InputType[_], in: Option[In])(implicit
+        um: InputUnmarshaller[In]): Vector[Violation] =
+      // early termination if errors limit is defined and the current number of violations exceeds the limit
+      if (errorsLimit.exists(_ <= errors)) Vector.empty
+      else
+        (tpe, in) match {
+          case (OptionInputType(ofType), Some(value)) if um.isDefined(value) =>
+            isValidValueRec(ofType, Some(value))
+          case (OptionInputType(_), _) => Vector.empty
+          case (_, None) => addViolation(NotNullValueIsNullViolation(sourceMapper, Nil))
+
+          case (ListInputType(ofType), Some(values)) if um.isListNode(values) =>
+            um.getListValue(values)
+              .toVector
+              .flatMap(v =>
+                isValidValueRec(
+                  ofType,
+                  v match {
+                    case opt: Option[In @unchecked] => opt
+                    case other => Option(other)
+                  }).map(ListValueViolation(0, _, sourceMapper, Nil)))
+
+          case (ListInputType(ofType), Some(value)) if um.isDefined(value) =>
+            isValidValueRec(
+              ofType,
+              value match {
+                case opt: Option[In @unchecked] => opt
+                case other => Option(other)
+              }).map(ListValueViolation(0, _, sourceMapper, Nil))
+
+          case (objTpe: InputObjectType[_], Some(valueMap)) if um.isMapNode(valueMap) =>
+            val unknownFields = um.getMapKeys(valueMap).toVector.collect {
+              case f if !objTpe.fieldsByName.contains(f) =>
+                addViolation(
+                  UnknownInputObjectFieldViolation(
+                    SchemaRenderer.renderTypeName(objTpe, true),
+                    f,
+                    sourceMapper,
+                    Nil)).head
+            }
 
-      fieldViolations ++ unknownFields
+            val fieldViolations =
+              objTpe.fields.toVector.flatMap(f =>
+                isValidValueRec(f.fieldType, um.getMapValue(valueMap, f.name))
+                  .map(MapValueViolation(f.name, _, sourceMapper, Nil)))
 
-    case (objTpe: InputObjectType[_], _) =>
-      Vector(
-        InputObjectIsOfWrongTypeMissingViolation(
-          SchemaRenderer.renderTypeName(objTpe, true),
-          sourceMapper,
-          Nil))
+            fieldViolations ++ unknownFields
 
-    case (scalar: ScalarType[_], Some(value)) if um.isScalarNode(value) =>
-      val coerced = um.getScalarValue(value) match {
-        case node: ast.Value => scalar.coerceInput(node)
-        case other => scalar.coerceUserInput(other)
-      }
+          case (objTpe: InputObjectType[_], _) =>
+            addViolation(
```

Comment:
So instead of returning Vector(violation), we call `addViolation(violation)` that mutates number of errors and returns the same Vector(violation)